man/tpm2_duplicate.1.md
% tpm2_duplicate(1) tpm2-tools | General Commands Manual

# NAME

tpm2_duplicate(1) -  Duplicates a loaded object so that it may be used in a
different hierarchy.

# SYNOPSIS

**tpm2_duplicate** [*OPTIONS*]

# DESCRIPTION

**tpm2_duplicate**(1) - This tool duplicates a loaded object so that it may be
used in a different hierarchy. The new parent key for the duplicate may be on
the same or different TPM or TPM_RH_NULL.

# OPTIONS

These options control the key importation process:

  * **-G**, **\--wrapper-algorithm**=_ALGORITHM_:

    The symmetric algorithm to be used for the inner wrapper if -U is not used.
    Supports:
    * aes - AES 128 in CFB mode.
    * null - none
    The key algorithm associated with the public parent if -U is used.

  * **-G**, **\--key-algorithm**=_ALGORITHM_:

    The key algorithm associated with the public parent if -U is used.

  * **-i**, **\--encryptionkey-in**=_FILE_:

    Specifies the filename of the symmetric key (128 bit data) to be used for
    the inner wrapper. Valid only when specified symmetric algorithm is not null

  * **-o**, **\--encryptionkey-out**=_FILE_:

    Specifies the filename to store the symmetric key (128 bit data) that was
    used for the inner wrapper. Valid only when specified symmetric algorithm is
    not null and \--input-key-file is not specified. The TPM generates the key
    in this case.

  * **-C**, **\--parent-context**=_OBJECT_:

    The parent key object.

  * **-U**, **\--parent-public**=_FILE_:

    Specifies the file path to the public key of the parent object on the
    destination TPM.  This should be a `TPM2B_PUBLIC` formatted file.
    This public key is used for the wrapping of a PEM or DER key
    which will be exported for a remote TPM.

  * **-k**, **\--private-key**=_FILE_:

    Specifies the file path to the external private key be encrypted for
    the remote TPM.  This should be a PEM format private key.

  * **-r**, **\--private**=_FILE_:

    Specifies the file path to save the private portion of the duplicated object.
    [protection details](common/protection-details.md)

  * **-u**, **\--public**=_FILE_:

    Specifies the file path to save the public portion of the duplicated object,
    if an external key is being duplicated.

  * **-s**, **\--encrypted-seed**=_FILE_:

    The file to save the encrypted seed of the duplicated object.

  * **-p**, **\--auth**=_AUTH_:

    The authorization value for the key, optional.

  * **-L**, **\--policy**=_FILE_:

    The input policy file, optional.

  * **-c**, **\--key-context**=_OBJECT_:

    The object to be duplicated.

  * **-a**, **\--attributes**=_ATTRIBUTES_:

    The object attributes, optional. The default for created objects is:

    `TPMA_OBJECT_SIGN_ENCRYPT|TPMA_OBJECT_DECRYPT|TPMA_OBJECT_USERWITHAUTH`

    Note: If a policy is specified without an auth value then `TPMA_OBJECT_USERWITHAUTH`
    is turned down.

  * **\--cphash**=_FILE_

    File path to record the hash of the command parameters. This is commonly
    termed as cpHash. NOTE: When this option is selected, The tool will not
    actually execute the command, it simply returns a cpHash.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[algorithm specifiers](common/alg.md) details the options for specifying
cryptographic algorithms _ALGORITHM_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

#### Example-1: Duplicate a key created with a policy that allows for duplication:

```bash
tpm2_startauthsession -S session.dat
tpm2_policycommandcode -S session.dat -L policy.dat TPM2_CC_Duplicate
tpm2_flushcontext session.dat

tpm2_createprimary -C o -g sha256 -G rsa -c primary.ctxt
tpm2_create -C primary.ctxt -g sha256 -G rsa -r key.prv -u key.pub  -c key.ctxt \
-L policy.dat -a "sensitivedataorigin|userwithauth|decrypt|sign"

tpm2_createprimary -C o -g sha256 -G ecc -c new_parent.ctxt

tpm2_startauthsession \--policy-session -S session.dat
tpm2_policycommandcode -S session.dat -L policy.dat TPM2_CC_Duplicate
tpm2_duplicate -C new_parent.ctxt -c key.ctxt -G null -p "session:session.dat" \
-r duprv.bin -s seed.dat
tpm2_flushcontext session.dat
```

#### Example-2: As an end-to-end example, the following will transfer an RSA key generated on `TPM-A` to `TPM-B`

##### On TPM-B

Create a parent object that will be used to wrap/transfer the key.
```
tpm2_createprimary -C o -g sha256 -G rsa -c primary.ctx

tpm2_create  -C primary.ctx -g sha256 -G rsa \
-r new_parent.prv  -u new_parent.pub \
-a "restricted|sensitivedataorigin|decrypt|userwithauth"
```

Copy `new_parent.pub` to `TPM-A`.

##### On TPM-A

Create root object and auth policy allows duplication only

```
tpm2_createprimary -C o -g sha256 -G rsa -c primary.ctx

tpm2_startauthsession -S session.dat

tpm2_policycommandcode -S session.dat -L dpolicy.dat TPM2_CC_Duplicate

tpm2_flushcontext session.dat

rm session.dat
```

Generate an RSA keypair on TPM-A that will be duplicated 
(note the passphrase is 'foo')

```
tpm2_create -C primary.ctx -g sha256 -G rsa -p foo -r key.prv \
-u key.pub  -L dpolicy.dat -a "sensitivedataorigin|userwithauth|decrypt|sign"

tpm2_load -C primary.ctx -r key.prv -u key.pub -c key.ctx

tpm2_readpublic -c key.ctx -o dup.pub
```

Test sign and encryption locally (so we can compare later that the same key 
was transferred).

```
echo "meet me at.." >file.txt
tpm2_rsaencrypt -c key.ctx  -o data.encrypted file.txt
tpm2_sign -c key.ctx -g sha256 -f plain -p foo -o sign.raw file.txt
```

Compare the signature hash (we will use this later to confirm the key was 
transferred to TPM-B):

```
sha256sum sign.raw

a1b4e3fbaa29e6e46d95cff498150b6b8e7d9fd21182622e8f5a3ddde257879e
```

Start an auth session and policy command to allow duplication
```
tpm2_startauthsession --policy-session -S session.dat

tpm2_policycommandcode -S session.dat -L dpolicy.dat TPM2_CC_Duplicate
```

Load the new_parent.pub file transferred from `TPM-B`
```
tpm2_loadexternal -C o -u new_parent.pub -c new_parent.ctx
```

Start the duplication
```
tpm2_duplicate -C new_parent.ctx -c key.ctx -G null  \
-p "session:session.dat" -r dup.dpriv -s dup.seed
```

Copy the following files to TPM-B:
* dup.pub
* dup.dpriv
* dup.seed
* (optionally data.encrypted just to test decryption)

##### On TPM-B

Start an auth,policy session
```
tpm2_startauthsession --policy-session -S session.dat

tpm2_policycommandcode -S session.dat -L dpolicy.dat TPM2_CC_Duplicate
```

Load the context we used to transfer
```
tpm2_flushcontext --transient-object

tpm2_load -C primary.ctx -u new_parent.pub -r new_parent.prv -c new_parent.ctx
```

Import the duplicated context against the parent we used
```
tpm2_import -C new_parent.ctx -u dup.pub -i dup.dpriv \
-r dup.prv -s dup.seed -L dpolicy.dat
```

Load the duplicated key context 
```
tpm2_flushcontext --transient-object

tpm2_load -C new_parent.ctx -u dup.pub -r dup.prv -c dup.ctx
```

Test the imported key matches

* Sign

```bash
echo "meet me at.." >file.txt

tpm2_sign -c dup.ctx -g sha256 -o sig.rss -p foo file.txt

dd if=sig.rss of=sign.raw bs=1 skip=6 count=256
```

Compare the signature file hash:

```bash
$ sha256sum sign.raw

a1b4e3fbaa29e6e46d95cff498150b6b8e7d9fd21182622e8f5a3ddde257879e
```

* Decryption

```
tpm2_flushcontext --transient-object

tpm2_rsadecrypt -p foo -c dup.ctx -o data.ptext data.encrypted

# cat data.ptext 
meet me at..
```

#### Example-3: Exporting an OpenSSL RSA key for a remote TPM

To securely send an OpenSSL generated RSA key to a remote TPM such that
only that remote TPM will be able to load it, and without exposing the
private key to the host operating system on the remote machine:

* On the destination TPM-B, create a primary context and read its public key,
then send `primary.pub` to the source machine:

```bash
tpm2_createprimary -c primary.ctx
tpm2_readpublic -c primary.ctx -o primary.pub
```

* On the source machine create the RSA private key and wrap it for the destination
TPM's public key.  Similar to `tpm2_makecredential`, this step should not require
a TPM.

```bash
openssl genrsa -out rsa.pem
tpm2_duplicate -U primary.pub -G rsa -k rsa.pem -u rsa.pub -r rsa.dpriv -s rsa.seed
```

* Send the `rsa.pub`, `rsa.dpriv` and `rsa.seed` to the destination TPM-B
and import the files, which will decrypt them using the `primary.ctx` to
produce `rsa.priv`, which can then be loaded and used as a TPM key:

```bash
tpm2_import -C primary.ctx -G rsa -i rsa.dpriv -s rsa.seed -u rsa.pub -r rsa.priv
tpm2_load -C primary.ctx -c rsa.ctx -u rsa.pub -r rsa.priv
```

#### Example-4: Exporting an HMAC key for a remote TPM and restrict it from further exports

This procedure will export an HMAC key created inside `TPM-A` only to `TPM-B` and with a 
policy that prevents `TPM-B` from allowing any further exports.

* On the destination `TPM-B`, create a primary context:

```bash
tpm2_createprimary -C o -g sha256 -G rsa -c primary.ctx
tpm2_create  -C primary.ctx -g sha256 -G rsa -r new_parent.prv \
    -u new_parent.pub -a "fixedtpm|fixedparent|restricted|sensitivedataorigin|decrypt|userwithauth"
```

Copy `new_parent.pub` to `TPM-A`.

* On source `TPM-A`

load the remote key and save its "name"

```bash
tpm2_loadexternal -C o -u new_parent.pub -c new_parent.ctx -n new_parent.name
```

Create a primary object and a policy that restricts duplication to just `TPM-B` parent

```bash
tpm2_createprimary -C o -g sha256 -G rsa -c primary.ctx
tpm2_startauthsession -S session.dat
tpm2_policyduplicationselect -S session.dat  -N new_parent.name -L dpolicy.dat 
tpm2_flushcontext session.dat
```

Create an HMAC key with that policy

```bash
tpm2_create -C primary.ctx -g sha256 -G hmac -r key.prv -u key.pub \
   -L dpolicy.dat -a "sensitivedataorigin|userwithauth|sign"
```

Load the key and save its "name".  Generate a test HMAC

```bash
tpm2_load -C primary.ctx -r key.prv -u key.pub -c key.ctx
tpm2_readpublic -c key.ctx -n dupkey.name

export plain="foo"

CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_duplicate.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include <tss2/tss2_mu.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_alg_util.h"
#include "tpm2_identity_util.h"
#include "tpm2_openssl.h"
#include "tpm2_options.h"
#include "tpm2_tool.h"

#define MAX_SESSIONS 3
typedef struct tpm_duplicate_ctx tpm_duplicate_ctx;
struct tpm_duplicate_ctx {
    /*
     * Inputs
     */
    /* common */
    struct {
        const char *ctx_path;
        const char *auth_str;
        char *policy_str;
        char *attr_str;
        tpm2_loaded_object object;
    } duplicable_key;

    char *key_type;

    /* tpm2_duplicate only */
    struct {
        const char *ctx_path;
        tpm2_loaded_object object;
    } new_parent_key;

    TPMT_SYM_DEF_OBJECT sym_alg;

    char *sym_key_in;
    TPM2B_DATA in_key;

    /* tpm2_openssl */
    const char *in_private_key_file;
    TPM2B_SENSITIVE in_private_key_data;
    const char *in_parent_public_key_file;
    TPM2B_PUBLIC in_parent_public_key_data;

    bool is_openssl_duplicate;

    /*
     * Outputs
     */
    /*common*/
    const char *out_duplicate_key_private_file;
    char *enc_seed_out;

    /* tpm2_openssl */
    const char *out_duplicate_key_public_file;
    TPM2B_PUBLIC out_public_data;

    /*tpm2_duplicate*/
    TPM2B_ENCRYPTED_SECRET *out_sym_seed;

    char *sym_key_out;
    TPM2B_DATA *out_key;

    TPM2B_PRIVATE *out_private_data;

    /*
     * Parameter hashes
     */
    const char *cp_hash_path;
    TPM2B_DIGEST cp_hash;
    bool is_command_dispatch;
    TPMI_ALG_HASH parameter_hash_algorithm;
};

static tpm_duplicate_ctx ctx = {
    .out_public_data = TPM2B_EMPTY_INIT,
    .in_private_key_data = TPM2B_EMPTY_INIT,
    .in_parent_public_key_data = TPM2B_EMPTY_INIT,
    .parameter_hash_algorithm = TPM2_ALG_ERROR,
};

static tool_rc duplicate(ESYS_CONTEXT *ectx) {

    return tpm2_duplicate(ectx, &ctx.duplicable_key.object,
        &ctx.new_parent_key.object, &ctx.in_key, &ctx.sym_alg, &ctx.out_key,
        &ctx.out_private_data, &ctx.out_sym_seed, &ctx.cp_hash,
        ctx.parameter_hash_algorithm);
}

static tool_rc openssl_create_duplicate(void) {

    /*
     * Calculate the object name.
     */
    tool_rc rc = tool_rc_success;
    TPM2B_NAME pubname = TPM2B_TYPE_INIT(TPM2B_NAME, name);
    bool result = tpm2_identity_create_name(&ctx.out_public_data, &pubname);
    if (!result) {
        rc = tool_rc_general_error;
        goto out;
    }

    TPM2B_DIGEST * seed = &ctx.in_private_key_data.sensitiveArea.seedValue;
    TPM2B_MAX_BUFFER hmac_key;
    TPM2B_MAX_BUFFER enc_key;
    tpm2_identity_util_calc_outer_integrity_hmac_key_and_dupsensitive_enc_key(
        &ctx.in_parent_public_key_data, &pubname, seed, &hmac_key, &enc_key);

    /*
     * Marshall the private key into a buffer
     */
    TPM2B_MAX_BUFFER marshalled_sensitive = TPM2B_EMPTY_INIT;
    size_t marshalled_sensitive_size = 0;
    TSS2_RC rval = Tss2_MU_TPMT_SENSITIVE_Marshal(
        &ctx.in_private_key_data.sensitiveArea,
        marshalled_sensitive.buffer + sizeof(marshalled_sensitive.size),
        TPM2_MAX_DIGEST_BUFFER, &marshalled_sensitive_size);
    if (rval != TPM2_RC_SUCCESS) {
        LOG_ERR("Error serializing sensitive area");
        rc = tool_rc_general_error;
        goto out;
    }

    size_t marshalled_sensitive_size_info = 0;
    rval = Tss2_MU_UINT16_Marshal(marshalled_sensitive_size,
        marshalled_sensitive.buffer, sizeof(marshalled_sensitive.size),
        &marshalled_sensitive_size_info);
    if (rval != TPM2_RC_SUCCESS) {
        LOG_ERR("Error serializing sensitive area size");
        rc = tool_rc_general_error;
        goto out;
    }

    marshalled_sensitive.size =
        marshalled_sensitive_size + marshalled_sensitive_size_info;

    /*
     * Compute the outer HMAC over the marshalled sensitive area
     * and encrypt it with the seed value.
     */
    TPM2B_DIGEST outer_hmac = TPM2B_EMPTY_INIT;
    TPM2B_MAX_BUFFER encrypted_duplicate_sensitive = TPM2B_EMPTY_INIT;
    tpm2_identity_util_calculate_outer_integrity(
        ctx.in_parent_public_key_data.publicArea.nameAlg,
        &pubname, &marshalled_sensitive, &hmac_key, &enc_key,
        &ctx.in_parent_public_key_data.publicArea.parameters.rsaDetail.symmetric,
        &encrypted_duplicate_sensitive, &outer_hmac);

    /*
     * Build the private data structure for writing out
     */
    TPM2B_PRIVATE private = TPM2B_EMPTY_INIT;
    UINT16 parent_hash_size = tpm2_alg_util_get_hash_size(
        ctx.in_parent_public_key_data.publicArea.nameAlg);

    private.size = sizeof(parent_hash_size) +
        parent_hash_size + encrypted_duplicate_sensitive.size;

    size_t hmac_size_offset = 0;
    rval = Tss2_MU_UINT16_Marshal(parent_hash_size, private.buffer,
        sizeof(parent_hash_size), &hmac_size_offset);
    if (rval != TPM2_RC_SUCCESS) {
        LOG_ERR("Error serializing hmac size");
        rc = tool_rc_general_error;
        goto out;
    }

    memcpy(private.buffer + hmac_size_offset, outer_hmac.buffer,
        parent_hash_size);
    memcpy(private.buffer + hmac_size_offset + parent_hash_size,
        encrypted_duplicate_sensitive.buffer,
        encrypted_duplicate_sensitive.size);

    ctx.out_private_data = malloc(private.size + sizeof(private.size));
    memcpy(ctx.out_private_data, &private, private.size + sizeof(private.size));

out:
    return rc;
}

#define DEFAULT_DUPLICATE_ATTRS (TPMA_OBJECT_USERWITHAUTH|TPMA_OBJECT_DECRYPT|TPMA_OBJECT_SIGN_ENCRYPT)
static void setup_default_attrs(TPMA_OBJECT *attrs, bool has_policy,
    bool has_auth) {

    /* Handle Default Setup */
    *attrs = DEFAULT_DUPLICATE_ATTRS;

    /*
     * IMPORTANT: if the object we're creating has a policy and NO authvalue, turn off userwith auth
     * so empty passwords don't work on the object.
     */
    if (has_policy && !has_auth) {
        *attrs &= ~TPMA_OBJECT_USERWITHAUTH;
    }
}

static tool_rc process_openssl_duplicate(void) {

    bool result = files_load_public(ctx.in_parent_public_key_file,
        &ctx.in_parent_public_key_data);
    if (!result) {
        return tool_rc_general_error;
    }

    TPMA_OBJECT attrs = 0;
    bool is_policy_specified = (ctx.duplicable_key.policy_str != 0);
    bool is_auth_specified = (ctx.duplicable_key.auth_str != 0);
    setup_default_attrs(&attrs, is_policy_specified, is_auth_specified);

    TPM2B_PUBLIC template = { 0 };
    tool_rc rc = tpm2_alg_util_public_init(ctx.key_type, 0, ctx.duplicable_key.attr_str,
        ctx.duplicable_key.policy_str, attrs, &template);
    if (rc != tool_rc_success) {
        return rc;
    }

    TPM2B_ENCRYPTED_SECRET encrypted_seed = TPM2B_EMPTY_INIT;
    result = tpm2_openssl_import_keys(&ctx.in_parent_public_key_data, &encrypted_seed,
        ctx.duplicable_key.auth_str, ctx.in_private_key_file, 0, &template,
        &ctx.in_private_key_data, &ctx.out_public_data);
    if (!result) {
        return tool_rc_general_error;
    }

    ctx.out_sym_seed = malloc(encrypted_seed.size +
        sizeof(encrypted_seed.size));
    memcpy(ctx.out_sym_seed, &encrypted_seed,
        encrypted_seed.size + sizeof(encrypted_seed.size));

    return tool_rc_success;
}

static bool set_key_algorithm(const char *algstr, TPMT_SYM_DEF_OBJECT * obj) {

    bool is_algstr_null = (strcmp(algstr, "null") == 0);
    if (is_algstr_null) {
        obj->algorithm = TPM2_ALG_NULL;
        return true;
    }

    bool is_algstr_aes = (strcmp(algstr, "aes") == 0);
    if (is_algstr_aes) {
        obj->algorithm = TPM2_ALG_AES;
        obj->keyBits.aes = 128;
        obj->mode.aes = TPM2_ALG_CFB;
        return true;
    }

    LOG_ERR("The algorithm \"%s\" is not supported!", algstr);
    return false;
}

static tool_rc process_output(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);
    /*
     * 1. Outputs that do not require TPM2_CC_<command> dispatch
     */
    bool is_file_op_success = true;
    if (ctx.cp_hash_path && !ctx.is_openssl_duplicate) {
        is_file_op_success = files_save_digest(&ctx.cp_hash, ctx.cp_hash_path);

        if (!is_file_op_success) {
            return tool_rc_general_error;
        }
    }

    tool_rc rc = tool_rc_success;
    if (!ctx.is_command_dispatch && !ctx.is_openssl_duplicate) {
        return rc;
    }

    /*
     * 2. Outputs generated after TPM2_CC_<command> dispatch
     */
    if (ctx.sym_key_out && !ctx.is_openssl_duplicate) {
        if (ctx.out_key == 0) {
            LOG_ERR("No encryption key from TPM ");
            rc = tool_rc_general_error;
            goto out;
        }
        is_file_op_success = files_save_bytes_to_file(ctx.sym_key_out,
            ctx.out_key->buffer, ctx.out_key->size);
        if (!is_file_op_success) {
            LOG_ERR("Failed to save encryption key out into file \"%s\"",
                    ctx.sym_key_out);
            rc = tool_rc_general_error;
            goto out;
        }
    }

    is_file_op_success = files_save_encrypted_seed(ctx.out_sym_seed, ctx.enc_seed_out);
    if (!is_file_op_success) {
        LOG_ERR("Failed to save encryption seed into file \"%s\"",
                ctx.enc_seed_out);
        rc = tool_rc_general_error;
        goto out;
    }

    is_file_op_success = files_save_private(ctx.out_private_data,
        ctx.out_duplicate_key_private_file);
    if (!is_file_op_success) {
        LOG_ERR("Failed to save private key into file \"%s\"",
                ctx.out_duplicate_key_private_file);
        rc = tool_rc_general_error;
        goto out;
    }

    if (ctx.is_openssl_duplicate) {
        is_file_op_success = files_save_public(&ctx.out_public_data,
            ctx.out_duplicate_key_public_file);
        if (!is_file_op_success) {
            LOG_ERR("Failed to save public key into file \"%s\"",
                ctx.out_duplicate_key_public_file);
            rc = tool_rc_general_error;
            goto out;
        }
    }

out:
    return rc;
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    /*
     * 3.a Command specific initializations
     *
     * No further processing as duplication is handled without going to the TPM
     * 
     */
    if (ctx.is_openssl_duplicate) {
        return process_openssl_duplicate();
    }

    /*
     * 1. Object and auth initializations
     */

    /*
     * 1.a Add the new-auth values to be set for the object.
     */

    /*
     * 1.b Add object names and their auth sessions
     */

    /* Object #1 */
    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.duplicable_key.ctx_path,
            ctx.duplicable_key.auth_str, &ctx.duplicable_key.object, false,
            TPM2_HANDLE_ALL_W_NV);
    if (rc != tool_rc_success) {
        LOG_ERR("Invalid authorization");
        return rc;
    }

    /* Object #2 */
    rc = tpm2_util_object_load(ectx, ctx.new_parent_key.ctx_path,
            &ctx.new_parent_key.object, TPM2_HANDLE_ALL_W_NV);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 2. Restore auxiliary sessions
     */

    /*
     * 3.b Command specific initializations
     */
    bool result = set_key_algorithm(ctx.key_type, &ctx.sym_alg);
    if (!result) {
        return tool_rc_general_error;
    }

    if (ctx.sym_key_in) {
        ctx.in_key.size = 16;
        result = files_load_bytes_from_path(ctx.sym_key_in, ctx.in_key.buffer,
                &ctx.in_key.size);
        if (!result) {
            return tool_rc_general_error;
        }
        if (ctx.in_key.size != 16) {
            LOG_ERR("Invalid AES key size, got %u bytes, expected 16",
                    ctx.in_key.size);
            return tool_rc_general_error;
        }
    }

    /*
     * 4. Configuration for calculating the pHash
     */

    /*
     * 4.a Determine pHash length and alg
     */
    tpm2_session *all_sessions[MAX_SESSIONS] = {
        ctx.duplicable_key.object.session,
        0,
        0
    };

    const char **cphash_path = ctx.cp_hash_path ? &ctx.cp_hash_path : 0;

    ctx.parameter_hash_algorithm = tpm2_util_calculate_phash_algorithm(ectx,
        cphash_path, &ctx.cp_hash, 0, 0, all_sessions);
    /*
     * 4.b Determine if TPM2_CC_<command> is to be dispatched
     */
    ctx.is_command_dispatch = ctx.cp_hash_path ? false : true;

    return rc;
}

static tool_rc check_options(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    bool result = true;
    /* Check for NULL alg & (keyin | keyout) */
    if (!ctx.key_type) {
        LOG_ERR("Expected key type to be specified via \"-G\","
                " missing option.");
        result = false;
    }

    /* If -G is not "null" we need an encryption key */
    bool is_key_type_not_null = strcmp(ctx.key_type, "null");
    bool is_in_or_out_enc_key_expected =
        (is_key_type_not_null && !ctx.in_parent_public_key_file);

    if (is_in_or_out_enc_key_expected) {
        if (!ctx.sym_key_in && !ctx.sym_key_out) {
            LOG_ERR("Expected in or out encryption key file \"-i/o\","
                    " missing option.");
            result = false;
        }

        if (ctx.sym_key_in && ctx.sym_key_out) {
            LOG_ERR("Expected either in or out encryption key file \"-i/o\","
                    " conflicting options.");
            result = false;
        }
    }

    if (!is_in_or_out_enc_key_expected && (ctx.sym_key_in || ctx.sym_key_out)) {
        LOG_ERR("Expected neither in nor out encryption key file \"-i/o\","
                " conflicting options.");
        result = false;
    }

    bool is_parent_public_specified = (ctx.in_parent_public_key_file != 0);
    bool is_local_private_specified = (ctx.in_private_key_file != 0);
    bool is_parent_public_and_local_private_availability_conflict =
        (is_parent_public_specified != is_local_private_specified);

    if (is_parent_public_and_local_private_availability_conflict) {
        LOG_ERR("Conflicting options: remote public key and local private key "
                "must both be specified");
        result = false;
    }

    bool is_neither_parent_public_nor_local_private_spcified =
        (!is_parent_public_specified && !is_local_private_specified);
    if (is_neither_parent_public_nor_local_private_spcified) {
	    if (!ctx.new_parent_key.ctx_path) {
	        LOG_ERR("Expected new parent object to be specified via \"-C\","
	        " missing option.");
	        result = false;
	    }

	    if (!ctx.duplicable_key.ctx_path) {
	        LOG_ERR("Expected object to be specified via \"-c\","
	        " missing option.");
	        result = false;
	    }

	    if (!ctx.enc_seed_out) {
	        LOG_ERR(
	        "Expected encrypted seed out filename to be specified via \"-S\","
	    	    " missing option.");
	        result = false;
	    }

	    if (!ctx.out_duplicate_key_private_file) {
	        LOG_ERR("Expected private key out filename to be specified via \"-r\","
	        " missing option.");
	        result = false;
	    }
    }

    return result ? tool_rc_success : tool_rc_option_error;
}

static bool on_option(char key, char *value) {

    switch (key) {
    case 'p':
        ctx.duplicable_key.auth_str = value;
        break;
    case 'L':
        ctx.duplicable_key.policy_str = value;
        break;
    case 'G':
        ctx.key_type = value;
        break;
    case 'i':
        ctx.sym_key_in = value;
        break;
    case 'o':
        ctx.sym_key_out = value;
        break;
    case 'C':
        ctx.new_parent_key.ctx_path = value;
        break;
    case 'c':
        ctx.duplicable_key.ctx_path = value;
        break;
    case 'r':
        ctx.out_duplicate_key_private_file = value;
        break;
    case 'u':
        ctx.out_duplicate_key_public_file = value;
        break;
    case 's':
        ctx.enc_seed_out = value;
        break;
    case 'U':
        ctx.in_parent_public_key_file = value;
        ctx.is_openssl_duplicate = true;
        break;
    case 'k':
        ctx.in_private_key_file = value;
        break;
    case 'a':
        ctx.duplicable_key.attr_str = value;
        break;
    case 0:
        ctx.cp_hash_path = value;
        break;
    default:
        LOG_ERR("Invalid option");
        return false;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    const struct option topts[] = {
      { "auth",              required_argument, 0, 'p'},
      { "policy",            required_argument, 0, 'L'},
      { "wrapper-algorithm", required_argument, 0, 'G'},
      { "key-algorithm",     required_argument, 0, 'G'},
      { "private",           required_argument, 0, 'r'},
      { "public",            required_argument, 0, 'u'},
      { "private-key",       required_argument, 0, 'k'},
      { "encryptionkey-in",  required_argument, 0, 'i'},
      { "encryptionkey-out", required_argument, 0, 'o'},
      { "encrypted-seed",    required_argument, 0, 's'},
      { "parent-context",    required_argument, 0, 'C'},
      { "parent-public",     required_argument, 0, 'U'},
      { "key-context",       required_argument, 0, 'c'},
      { "attributes",        required_argument, 0, 'a'},
      { "cphash",            required_argument, 0,  0 },
    };

    *opts = tpm2_options_new("p:L:G:i:C:o:s:r:c:U:k:u:a:", ARRAY_LEN(topts), topts,
            on_option, 0, TPM2_OPTIONS_OPTIONAL_SAPI);

    return *opts != 0;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    /*
     * 1. Process options
     */
    tool_rc rc = check_options(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 2. Process inputs
     */
    rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 3. TPM2_CC_<command> call
     */
    rc = ctx.is_openssl_duplicate ? openssl_create_duplicate() : duplicate(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 4. Process outputs
     */
    return process_output(ectx);
}

static tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Free objects
     */
    free(ctx.out_key);
    free(ctx.out_sym_seed);
    free(ctx.out_private_data);
    /*
     * 2. Close authorization sessions
     */
    tool_rc rc = tpm2_session_close(&ctx.duplicable_key.object.session);

    /*
     * 3. Close auxiliary sessions
     */

    return rc;
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("duplicate", tpm2_tool_onstart, tpm2_tool_onrun,
    tpm2_tool_onstop, 0)
